# Pandas Exercises
Practice exercises based on the employees.csv dataset.

## Setup

In [59]:
import pandas as pd

path_csv = "files/employees.csv"
df = pd.read_csv(path_csv)
df.head()

,id,name,age,department,role,city,salary,years_at_company,performance_rating,remote_work
0,1,Uma Brown,39,Engineering,Senior Engineer,Chicago,NaN,18,1.3,No
1,2,Henry Martin,23,Support,Specialist,Atlanta,99543.18,8,NaN,No
2,3,Zach Garcia,43,IT,Analyst,Austin,66955.20,4,1.4,Yes
3,4,Iris Johnson,56,Legal,Senior Engineer,Miami,49355.01,18,2.2,Hybrid
4,5,Sam Martinez,24,Marketing,Senior Engineer,Seattle,111045.28,8,4.5,No


---
## Section 1 — Exploring the DataFrame

**Exercise 1.1** — Display the first 5 rows of the DataFrame.

In [60]:
# your code here
df.head()


,id,name,age,department,role,city,salary,years_at_company,performance_rating,remote_work
0,1,Uma Brown,39,Engineering,Senior Engineer,Chicago,NaN,18,1.3,No
1,2,Henry Martin,23,Support,Specialist,Atlanta,99543.18,8,NaN,No
2,3,Zach Garcia,43,IT,Analyst,Austin,66955.20,4,1.4,Yes
3,4,Iris Johnson,56,Legal,Senior Engineer,Miami,49355.01,18,2.2,Hybrid
4,5,Sam Martinez,24,Marketing,Senior Engineer,Seattle,111045.28,8,4.5,No


**Exercise 1.2** — Display the last 8 rows.

In [61]:
# your code here
df.tail(8)

,id,name,age,department,role,city,salary,years_at_company,performance_rating,remote_work
292,293,Tina Davis,55,Engineering,Consultant,Boston,89968.73,7,4.2,No
293,294,Brian Anderson,23,Engineering,Senior Engineer,Chicago,51986.88,17,4.4,Yes
294,295,Mia White,43,Marketing,Team Lead,Chicago,69702.99,14,4.2,No
295,296,Rachel Harris,36,Operations,Specialist,Chicago,111536.14,10,3.9,No
296,297,Wendy White,58,IT,Analyst,Dallas,38724.02,20,1.9,Hybrid
297,298,Karen Williams,44,HR,Senior Engineer,Chicago,147973.66,11,2.8,No
298,299,Leo Lopez,22,Marketing,Specialist,Denver,40536.96,6,4.5,No
299,300,Mia Brown,58,Legal,Team Lead,Austin,109861.50,15,1.3,Yes


**Exercise 1.3** — Check column names, data types, and non-null counts.

In [62]:
# your code here
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   id                  300 non-null    int64  
 1   name                300 non-null    object 
 2   age                 300 non-null    int64  
 3   department          300 non-null    object 
 4   role                300 non-null    object 
 5   city                300 non-null    object 
 6   salary              282 non-null    float64
 7   years_at_company    300 non-null    int64  
 8   performance_rating  290 non-null    float64
 9   remote_work         300 non-null    object 
dtypes: float64(2), int64(3), object(5)
memory usage: 23.6+ KB


**Exercise 1.4** — Get a statistical summary of all numeric columns.

In [63]:
# your code here
df.describe()

,id,age,salary,years_at_company,performance_rating
count,300.000000,300.000000,282.000000,300.000000,290.000000
mean,150.500000,41.906667,86787.143085,10.490000,2.947931
std,86.746758,11.750669,32237.958066,5.814573,1.137173
min,1.000000,22.000000,35678.050000,1.000000,1.000000
25%,75.750000,33.000000,57273.662500,5.000000,2.000000
50%,150.500000,42.000000,84823.820000,10.000000,2.900000
75%,225.250000,53.000000,113647.217500,16.000000,3.975000
max,300.000000,60.000000,149873.770000,20.000000,5.000000


**Exercise 1.5** — How many rows and columns does the DataFrame have?

In [64]:
# your code here
df.shape #rows x columns

(300, 10)

---
## Section 2 — Selecting Columns

**Exercise 2.1** — Select only the `name` and `department` columns.

In [65]:
# your code here
df[["name", "department"]]

,name,department
0,Uma Brown,Engineering
1,Henry Martin,Support
2,Zach Garcia,IT
3,Iris Johnson,Legal
4,Sam Martinez,Marketing
...,...,...
295,Rachel Harris,Operations
296,Wendy White,IT
297,Karen Williams,HR
298,Leo Lopez,Marketing


**Exercise 2.2** — Select only the `name`, `role`, `salary`, and `city` columns.


In [66]:
# your code here
df[["name","role","salary","city"]]

,name,role,salary,city
0,Uma Brown,Senior Engineer,NaN,Chicago
1,Henry Martin,Specialist,99543.18,Atlanta
2,Zach Garcia,Analyst,66955.20,Austin
3,Iris Johnson,Senior Engineer,49355.01,Miami
4,Sam Martinez,Senior Engineer,111045.28,Seattle
...,...,...,...,...
295,Rachel Harris,Specialist,111536.14,Chicago
296,Wendy White,Analyst,38724.02,Dallas
297,Karen Williams,Senior Engineer,147973.66,Chicago
298,Leo Lopez,Specialist,40536.96,Denver


**Exercise 2.3** — Create a Series with all values from the `salary` column.


In [67]:
# your code here
series = df["salary"]


---
## Section 3 — Filtering

**Exercise 3.1** — Filter employees with salary greater than 100,000.

In [68]:
# your code here
salary_gt = df["salary"] > 100000
df[salary_gt]

,id,name,age,department,role,city,salary,years_at_company,performance_rating,remote_work
4,5,Sam Martinez,24,Marketing,Senior Engineer,Seattle,111045.28,8,4.5,No
6,7,Mia Lopez,36,Design,Manager,New York,113730.64,2,4.2,No
13,14,Sam Williams,53,Marketing,Team Lead,Atlanta,128842.67,5,3.6,Hybrid
14,15,Tina Hernandez,56,HR,Director,Austin,121855.50,13,5.0,Hybrid
16,17,Bob Davis,24,Marketing,Associate,San Francisco,133858.07,8,2.1,No
...,...,...,...,...,...,...,...,...,...,...
287,288,Grace Martin,56,HR,Analyst,Miami,126407.32,16,3.2,No
290,291,Henry White,44,IT,Team Lead,Chicago,142548.69,8,3.4,Hybrid
295,296,Rachel Harris,36,Operations,Specialist,Chicago,111536.14,10,3.9,No
297,298,Karen Williams,44,HR,Senior Engineer,Chicago,147973.66,11,2.8,No


**Exercise 3.2** — Filter employees who work in the `Engineering` department.


In [69]:
# your code here
department_filter = df["department"] == "Engineering"
df[department_filter]

,id,name,age,department,role,city,salary,years_at_company,performance_rating,remote_work
0,1,Uma Brown,39,Engineering,Senior Engineer,Chicago,NaN,18,1.3,No
12,13,Yara Martin,60,Engineering,Senior Engineer,New York,72277.95,12,4.5,No
25,26,Derek Williams,51,Engineering,Senior Engineer,Dallas,106433.31,3,3.2,Hybrid
51,52,Olivia Garcia,38,Engineering,Coordinator,Austin,78596.94,11,2.4,No
69,70,Carol Garcia,48,Engineering,Coordinator,Dallas,86807.93,10,1.1,No
77,78,Uma Moore,35,Engineering,Director,Chicago,65711.47,15,4.5,No
101,102,Uma Harris,25,Engineering,Team Lead,Austin,125481.38,19,1.8,Hybrid
111,112,Yara Hernandez,40,Engineering,Senior Engineer,Boston,104047.16,19,3.0,Yes
113,114,Tina Lee,60,Engineering,Junior Engineer,Seattle,141844.69,6,2.1,No
128,129,Uma Garcia,37,Engineering,Associate,Denver,116376.50,17,3.4,No


**Exercise 3.3** — Filter employees older than 40 AND with salary above 80,000.

In [70]:
# your code here
age_filter = df["age"] > 40
salary_filter = df["salary"] > 80000
df[age_filter&salary_filter]

,id,name,age,department,role,city,salary,years_at_company,performance_rating,remote_work
7,8,Sam Anderson,53,HR,Team Lead,Denver,80497.67,9,1.6,Hybrid
13,14,Sam Williams,53,Marketing,Team Lead,Atlanta,128842.67,5,3.6,Hybrid
14,15,Tina Hernandez,56,HR,Director,Austin,121855.50,13,5.0,Hybrid
20,21,Derek Davis,48,Sales,Specialist,Austin,90848.73,2,1.7,Yes
21,22,Zach Moore,49,Finance,Coordinator,Atlanta,115108.02,5,1.8,Yes
...,...,...,...,...,...,...,...,...,...,...
287,288,Grace Martin,56,HR,Analyst,Miami,126407.32,16,3.2,No
290,291,Henry White,44,IT,Team Lead,Chicago,142548.69,8,3.4,Hybrid
292,293,Tina Davis,55,Engineering,Consultant,Boston,89968.73,7,4.2,No
297,298,Karen Williams,44,HR,Senior Engineer,Chicago,147973.66,11,2.8,No


**Exercise 3.4** — Filter employees who work remotely (`remote_work == 'Yes'`) OR have a performance rating above 4.5.

In [71]:
# your code here
remote_work = df["remote_work"] == "Yes"
performance_rating = df["performance_rating"] > 4.5

df[remote_work & performance_rating].shape

(8, 10)

**Exercise 3.5** — Filter employees from `New York` who have been at the company for more than 10 years.


In [72]:
# your code here
city_filter = df["city"] == "New York"
years_at_company_filter = df["years_at_company"] > 10

df[city_filter & years_at_company_filter]


,id,name,age,department,role,city,salary,years_at_company,performance_rating,remote_work
12,13,Yara Martin,60,Engineering,Senior Engineer,New York,72277.95,12,4.5,No
15,16,Olivia Brown,36,HR,Consultant,New York,NaN,18,1.9,Yes
28,29,Victor Lopez,53,Design,Senior Engineer,New York,NaN,14,4.3,Yes
47,48,Noah Hernandez,56,Operations,Manager,New York,88581.29,14,NaN,Hybrid
73,74,Alice Williams,58,HR,Director,New York,102519.14,19,1.2,Yes
83,84,Wendy Thomas,31,IT,Team Lead,New York,91962.79,17,4.9,No
91,92,Victor Thomas,22,Finance,Consultant,New York,100081.84,16,4.3,No
105,106,Xander Martin,51,Finance,Junior Engineer,New York,108984.18,16,4.4,No
114,115,Frank Jones,47,Support,Senior Engineer,New York,43002.32,17,1.9,No
136,137,Frank Martinez,30,Support,Coordinator,New York,135395.22,12,3.2,Yes


**Exercise 3.6** — Check which rows have `salary > 120000` — return a boolean Series.

In [73]:
# your code here
df["salary"] > 120000

0      False
1      False
2      False
3      False
4      False
       ...  
295    False
296    False
297     True
298    False
299    False
Name: salary, Length: 300, dtype: bool

---
## Section 4 — Aggregations

**Exercise 4.1** — What is the total salary of all employees?

In [74]:
# your code here
df["salary"].sum()

np.float64(24473974.349999998)

**Exercise 4.2** — What is the average salary?


In [75]:
# your code here
df["salary"].mean()

np.float64(86787.14308510638)

**Exercise 4.3** — What is the maximum salary? And the minimum?


In [76]:
# your code here
df["salary"].min()
df["salary"].max()

np.float64(149873.77)

**Exercise 4.4** — What is the average `performance_rating` of employees?


In [77]:
# your code here
df["performance_rating"].mean()

np.float64(2.947931034482759)

**Exercise 4.5** — What is the average number of years employees have been at the company?


In [ ]:
# your code here
df["years_at_company"].mean()

---
## Section 5 — Adding and Removing Columns

**Exercise 5.1** — Add a new column `salary_with_bonus` equal to `salary * 1.10` (10% bonus).


In [79]:
# your code here


**Exercise 5.2** — Add a column `seniority` with value `'Senior'` if `years_at_company >= 10`, else `'Junior'`.

In [80]:
import numpy as np
# your code here


**Exercise 5.3** — Remove the `salary_with_bonus` column you just created.

In [81]:
# your code here


**Exercise 5.4** — Remove both `seniority` and `id` columns at once.

In [82]:
# your code here


---
## Section 6 — Handling Missing Values

**Exercise 6.1** — How many missing values are there in each column?

In [83]:
# your code here


**Exercise 6.2** — Remove all rows that have ANY missing value.

In [84]:
# your code here


**Exercise 6.3** — Remove rows that have missing values ONLY in the `salary` column.

In [85]:
# your code here


**Exercise 6.4** — Fill missing `salary` values with the column mean.

In [86]:
# your code here


**Exercise 6.5** — Fill missing `performance_rating` values with the median.


In [87]:
# your code here


**Exercise 6.6** — Convert the `salary` column to integer type after filling NaN values.

In [88]:
# your code here


---
## Section 7 — Value Counts & GroupBy

**Exercise 7.1** — Count how many employees are in each department.

In [89]:
# your code here


**Exercise 7.2** — Count how many employees have each type of `remote_work` arrangement.


In [90]:
# your code here


**Exercise 7.3** — Calculate the average salary per department.

In [91]:
# your code here


**Exercise 7.4** — Calculate the total salary paid per city.


In [92]:
# your code here


**Exercise 7.5** — Find the maximum `performance_rating` per department.


In [93]:
# your code here


**Exercise 7.6** — Group by `department` and `remote_work`, then calculate the average salary for each combination.


In [94]:
# your code here


---
## Section 8 — Sorting

**Exercise 8.1** — Sort employees by salary in ascending order.

In [95]:
# your code here


**Exercise 8.2** — Sort employees by salary in descending order.


In [96]:
# your code here


**Exercise 8.3** — Sort by `department` (A→Z), then by `salary` (highest first) within each department.

In [97]:
# your code here


**Exercise 8.4** — Find the top 5 highest-paid employees.

In [98]:
# your code here


---
## Section 9 — Merge & Concat

**Exercise 9.1** — Create two DataFrames: one with employees from `Engineering`, another from `Marketing`. Then concatenate them into a single DataFrame.

In [99]:
# your code here


**Exercise 9.2** — Create a small DataFrame with department budget info (make it up), then merge it with the main `df` on the `department` column.

Example budget DataFrame:
```python
budgets = pd.DataFrame({
    'department': ['Engineering', 'Marketing', 'Sales', 'HR', 'Finance'],
    'annual_budget': [5_000_000, 2_000_000, 3_500_000, 1_000_000, 2_500_000]
})
```

In [100]:
# your code here


**Exercise 9.3** — After the merge from 9.2, check the shape of the new DataFrame vs the original. How many rows were kept?


In [101]:
# your code here


---
## Section 10 — Challenge Exercises

**Challenge 10.1** — Find the department with the highest average salary.

In [102]:
# your code here


**Challenge 10.2** — How many employees earn above the overall average salary? What percentage is that?


In [103]:
# your code here


**Challenge 10.3** — Create a summary table showing, for each `city`:
- Total number of employees
- Average salary
- Average performance rating

In [104]:
# your code here


**Challenge 10.4** — Find all employees who are both in the top 10% salary AND have a performance rating above 4.0.

In [105]:
# your code here


**Challenge 10.5** — Rank all employees by salary within their department (1 = highest paid in that dept).

In [106]:
# your code here
